### ISSUE TITLE
[BUG] Long time for KerasSequenceLoader to "reload" after each epoch 

### ISSUE THREAD

**Issue by nv-dmajchrowski** [NVIDIA-Merlin/NVTabular#1142]
**Describe the bug**
I'm measuring time for each batch to be loaded using KerasSequenceLoader and I'm experiencing long time to load the first batch after previous epoch ends. For example: 

``` 
...
Epoch 0 Iter 455 0.001s
Epoch 0 Iter 456 0.003s
Epoch 1 Iter 000 1.162s
Epoch 1 Iter 001 0.001s
Epoch 1 Iter 002 0.001s
...
Epoch 1 Iter 455 0.001s
Epoch 1 Iter 456 0.003s
Epoch 2 Iter 000 1.570s
Epoch 2 Iter 001 0.001s
Epoch 2 Iter 002 0.001s
```

**Steps/Code to reproduce bug**

1. Use NVTabular/examples/advanced-ops-outbrain/ jupyter notebooks as an example
2. In last notebook (03-Training) add following cell (right after `train_dataset_tf` definition):

```
import time
for epoch in range(3):
    t = time.time()
    for it, _ in enumerate(train_dataset_tf):
        time.sleep(0.001)
        it_padded = str(it).zfill(3)
        print("Epoch", epoch, "Iter", f"{it_padded}", f"{time.time() - t:.3f}s")
        t = time.time()
```

**Expected behavior**
No overhead from data loading when starting new epoch

**Environment details (please complete the following information):**
 - Environment location: Docker
 - Method of NVTabular install:
   - Merlin tensorflow container - [code](https://github.com/NVIDIA-Merlin/Merlin/blob/f5fb5d5b9ad12c3cfc3c879e379d2e4d1f212daf/docker/dockerfile.tf)
   - Checkout to nvtabular v0.6.1 - `RUN cd /nvtabular && git pull && git checkout v0.6.1`
 

**Additional context**
- Measured on DGX1V 32GB

**Comment by rjzamora** [comment]
On my local machine, I am finding that about 50% of the delay corresponds to [this call to tf.split](https://github.com/NVIDIA/NVTabular/blob/2fc023356ec0d47d19a5c3a891ac294b18ff65ce/nvtabular/loader/tensorflow.py#L311) (which is called within [https://github.com/NVIDIA/NVTabular/blob/2fc023356ec0d47d19a5c3a891ac294b18ff65ce/nvtabular/loader/backend.py#L350](https://github.com/NVIDIA/NVTabular/blob/2fc023356ec0d47d19a5c3a891ac294b18ff65ce/nvtabular/loader/backend.py#L350) when a new chunk is loaded).

**Comment by rjzamora** [comment]
Okay - I am begining to understand what this issue is getting at, and I may have a partial solution.  I was getting caught up on something slightly different.

The DataLoader logic does use pre-fetching to read the next data partition into memory while the previous partition is being fed into the model in batches. However, this prefetching stops when the end of the underlying DataFrameIter is reached.  Therefore, when a multi-epoch training session is implemented as distinct loop over dataloader passes, each epoch will begin with a blocking data-fetch operation. The simplest solution to this problem is to add an explicit `epochs=` argument to the `DataLoader`.  This way, we could do something like `train_dataset_tf = KerasSequenceLoader(..., epochs=3)` to define a dataloader that will pass over the dataset three times, without loosing inter-epoch pre-fetching.

When I add this argument for the outbrains example, I only get a performance dip for the very first iteration:

```python
t = time.time()
for it, _ in enumerate(train_dataset_tf):
    time.sleep(0.02)
    tstep = time.time() - t
    if tstep > 0.4:
        it_padded = str(it).zfill(3)
        print("Multi-Epoch -- Iter", f"{it_padded}", f"{tstep:.3f}s")
    t = time.time()
print(f"Total Iterations: {str(it + 1).zfill(3)}")
```
```
Multi-Epoch -- Iter 000 1.428s
Total Iterations: 1368
```

With `main`, I get a performance dip for every new epoch as well:

```python
for epoch in range(3):
    t = time.time()
    for it, _ in enumerate(train_dataset_tf):
        time.sleep(0.02)
        tstep = time.time() - t
        if tstep > 0.4:
            it_padded = str(it).zfill(3)
            print("Epoch", epoch, "Iter", f"{it_padded}", f"{tstep:.3f}s")
        t = time.time()
print(f"Total Iterations: {str((it + 1) * 3).zfill(3)}")
```
```
Epoch 0 Iter 000 1.407s
Epoch 1 Iter 000 1.320s
Epoch 2 Iter 000 1.387s
Total Iterations: 1368
```

@benfred - Does this sound like a reasonable API change?

**Comment by rjzamora** [comment]
cc @jperez999 as well

**Comment by nv-dmajchrowski** [comment]
```
Just to clarify: Are you labeling this as a "bug" because earlier versions of NVTabular do not show any data-loader delay at the beginning of a new epoch, or because the delay is now larger?
```
Honestly I didn't know which category should it belong to. I was expecting the delay to be much smaller and wanted to know the root cause and if it's possible to do something about that. If it's more like a new feature (the `epochs`  argument) not a bug I'm fine with changing the label here.

**Comment by rjzamora** [comment]
>Honestly I didn't know which category should it belong to. I was expecting the delay to be much smaller and wanted to know the root cause

Cool - That makes sense. I'm not too concerned with the specific github label, but I did want to better understand the context of the issue.

**Comment by jperez999** [comment]
IMO, if we set that epochs argument that it locks the data loader to using the number of epochs on iteration...? Meaning if I set epochs=3 that data loader will ALWAYS iterate 3x over the data? Seems a little dangerous and counter intuitiveIs this really an issue...? @EvenOldridge.

**Comment by EvenOldridge** [comment]
I agree the dataloader should probably not need to explicitly know the number of epochs.  

Is it possible to prebuffer automatically on the assumption that there will be another epoch, with the final prebuffering being superfluous?  The cost is negligible from a perf perspective to do so, and then the user doesn't need to specify anything.

**Comment by karlhigley** [comment]
@jperez999 Yeah, I think this is really an issue, due to the perf drops between epochs.

@EvenOldridge I think the challenge (and correct me if I'm wrong @rjzamora) is that the `ChunkQueue` that backs the `DataLoader` stores a single iterator created by converting the `Dataset` object to an iterator over the underlying `DataFrame`, and the pre-buffering depends on using that same iterator throughout. Without knowing how many epochs we want to iterate through, the iterator can't properly compute its own size or elements.

**Comment by rjzamora** [comment]
I'm sure we can take some time to thing through a reasonable solution that will allow pre-fetching between epochs, but I am not yet convinced that it is worth the effort and added complexity. It may be the case that we can find an elegant solution that does not add a maintenence overhead or hurt code readability. However, it seems to me like most users are not going to care about a small data-reload overhead when they are using explicit python loops for mutli-epoch training workflows.  Users who want a highly optimized training pipeline can just bake a larger number of epochs into the iterator to get better performance when they really want/need it.

With my hesitation in mind - I am certainly happy to help work out a solution that will allow users to write "pythonic" for loops.

**Comment by karlhigley** [comment]
Yeah, given the set of constraints and trade-offs, I like #1147. It's simple and clean, most people will never know it's there, but anyone who needs it can take advantage.

**Comment by rjzamora** [comment]
Note that some variation on #1147 may also make sense. For example, we may want to allow the user to iterate over some `dataloader` obeject like: `for data in dataloader(epochs=3): ...`.  This way, the epoch count does not live on the dataloader object itself, but on the new iterator that is returned by `dataloader(epochs=3)`.

**Comment by karlhigley** [comment]
That might actually work okay, since neither of the underlying TF or Torch dataloader classes define a `__call__` method. 🤔

**Comment by rjzamora** [comment]
>That might actually work okay, since neither of the underlying TF or Torch dataloader classes define a __call__ method. 🤔

Right - I can update #1147 with a prototype. However, I am realizing that it would probably be best to add a dedicated method to produce a new multi-epoch dataloader (rather than using `__call__`), and not allow the user to directly initalize a multi-epoch dataloader with `KerasSequenceLoader` etc (to help avoid user error).

For example:

```python
# Start with general "single-epoch" dataloader
dataloader = Dataloader(dataset, ...)

# "Optimized" loop over three epochs:
for chunk in dataloader.<new-method>(3):
    ...
````
I am open to feedback on the best name for `<new-method>`. Does `dataloader.epochs` make sense, or should the name be more descriptive (e.g. `epoch_batch` or `continuous_epochs`)?

**Comment by karlhigley** [comment]
I dunno, I like `epochs`. Short and sweet, gets right to the point. "Give me a dataloader for three epochs." 👍🏻

**Comment by EvenOldridge** [comment]
Okay, epochs it is.  It should be fine, but maybe we can add something in the docs explaining why it's necessary.

**Comment by karlhigley** [comment]
Yeah, it should definitely be documented, but it's an optional parameter that gives you the most fully optimized behavior (if I understand correctly), so not totally necessary.
